File: go/array/169_majority_ele.go
package array

import "sort"

func majorityElement(nums []int) int {
	// Hash 记录每个元素出现的次数，如果某个元素出现次数大于n/2 则为expected
	// 时间复杂度O(n) + O(n) = 2 O(n), 空间复杂度O(n) 不是O(1)
	var hmap = make(map[int]int)

	for _, v := range nums {
		hmap[v] += 1
	}

	for k, v := range hmap {
		if v > len(nums)>>1 {
			return k
		}
	}
	panic("no majority elements")
}

func majorityElement2(nums []int) int {
	// 排序，众数一定位于n/2 的位置
	sort.Ints(nums)
	return nums[len(nums)/2]
}

// 摩尔投票法（Boyer–Moore majority vote algorithm) 如何在任意多的候选人中（选票无序），选出获得票数最多的那个。
// 算法可以分为两个阶段：
// 对抗阶段：分属两个候选人的票数进行两两对抗抵消
// 计数阶段：计算对抗结果中最后留下的候选人票数是否有效
func majorityElement3(nums []int) int {
	// major 被计票的元素, count 票数
	var major, count int

	// 	当我们遍历下一个选票时，判断当前 count 是否为零：
	//  若 count == 0，代表当前 major 空缺，直接将当前候选人赋值给 major，并令 cousort
	//  若 count != 0，代表当前 major 的票数未被完全抵消，因此令 count--，
	//  即使用当前候选人的票数抵消 major 的票数

	for _, v := range nums {
		if count == 0 {
			major = v
		}
		if major != v {
			count--
		} else {
			count++
		}
	}
	return major
}

File: rust/array/169_majority_ele.rs



pub fn majority_element(nums: Vec<i32>) -> i32 {
    // 排序
    // let mut nums = nums;
    // nums.sort();
    // nums[nums.len()/2]

    // 摩尔投票法 -- 用Option 枚举
    // let mut curr_elem = None;
    // let mut curr_count = 0;

    // for n in nums {
    //     match curr_elem {
    //         // 如果找到多数元素，对比新元素，继续计数
    //         Some(last) => {
    //             if last == n {
    //                 curr_count += 1;
    //             }else {
    //                 curr_count -= 1;
    //                 if curr_count == 0 {
    //                     curr_elem = None;
    //                 }
    //             }
    //         },
    //         None => {
    //             curr_elem = Some(n);
    //             curr_count += 1; 
    //         }
    //     }
    // }

    // match curr_elem {
    //     Some(majority) => majority,
    //     None => panic!("no majoriy")
    // }

     // 摩尔投票法 -- 不用枚举
    //  let mut marjor = 0;
    //  let mut count = 0;
    //  for n in nums {
    //     if count == 0 {
    //         marjor = n;
    //     }

    //     if marjor == n {
    //         count += 1;
    //     }else {
    //         count -= 1;
    //     }
    //  }
    //  marjor

    // Hmap
    use std::collections::HashMap;
    let mut hmap = HashMap::new();
    let ret = 0;

    for &n in &nums {
        let count = hmap.entry(n).or_insert(0);
        *count += 1;
    }

    for (k, v) in hmap {
        if v > (nums.len() / 2){
            return k
        }
    }
    ret
}
